dht/i_partitioner.cc

```diff
@@ -439,4 +443,44 @@ future<dht::partition_range_vector> subtract_ranges(const schema& schema, const
     co_return res;
 }
 
+dht::token_range_vector split_token_range_msb(unsigned most_significant_bits) {
+    dht::token_range_vector ret;
+    // Avoid shift left 64
+    if (!most_significant_bits) {
+        auto&& start_bound = dht::token_range::bound(dht::minimum_token(), true);
+        auto&& end_bound = dht::token_range::bound(dht::maximum_token(), true);
+        ret.emplace_back(std::move(start_bound), std::move(end_bound));
+        return ret;
+    }
+    uint64_t number_of_ranges = 1 << most_significant_bits;
+    ret.reserve(number_of_ranges);
+    assert(most_significant_bits < 64);
+    uint8_t log2_shift = (64 - most_significant_bits);
+    uint64_t keys_per_range = 1ULL << log2_shift;
+    uint64_t unbiased_key = 0;
+    dht::token token;
+    for (uint64_t i = 0; i < number_of_ranges; ) {
+        std::optional<dht::token_range::bound> start_bound;
+        std::optional<dht::token_range::bound> end_bound;
+        if (i == 0) {
+            start_bound = dht::token_range::bound(dht::minimum_token(), true);
+        } else {
+            start_bound = dht::token_range::bound(token, true);
+        }
+        if (++i < number_of_ranges) {
+            unbiased_key = i << log2_shift;
+            token = dht::bias(unbiased_key);
```

Comment:
I think I found a bug. 

end bound of group i should be (i << log_shift - 1), so its end bound don't overlap with start bound of group i + 1.

the check below failed to capture it because it checked the end bound using idx of next group (because *i* was incremented previously).

Reply:
the end bound for group i excludes the key while the atart bound for group i+1 includes it, so they don't overlap

Reply:
ah. missed that the end bound is exclusive if not the last range. it's very correct then.